### PR TITLE
BUG: Don't produce undefined behavior for a << b if b >= bitsof(a)

### LINKED ISSUE #2449: Shift operator cycles for int32 and int64 (Trac #1856)
_Original ticket http://projects.scipy.org/numpy/ticket/1856 on 2011-06-02 by trac user klasj, assigned to unknown._

The bitshift operations, both right and left cycles when they are shifted more bits than
the length of the datatype. This happens for both int32 and int64 but not for int16,
unsigned integers works as expected. In the example below version 1.6.1 is used but
this is also present in at least 1.5.0.
```
In [1]: import numpy as np

In [2]: np.right_shift(np.int32(8),32)

Out[2]: 8

In [3]: np.right_shift(np.int64(8),64)

Out[3]: 8

In [4]: np.right_shift(np.int16(8),16)

Out[4]: 0

In [5]: np.left_shift(np.int32(8),32)

Out[5]: 8

In [6]: np.left_shift(np.int64(8),64)

Out[6]: 8

In [7]: np.left_shift(np.int16(8),16)

Out[7]: 524288

In [8]: np.version.full_version

Out[8]: '1.6.1.dev-08953ea'
```

### PR DESCRIPTION
**Pull request by eric-wieser**
This is a continuation of gh-7473. Fixes gh-2449

### CODE REVIEW

**Comment by eric-wieser** [comment]
CI failure is:
```
numpy/core/src/umath/loops.c.src: In function ‘LONG_right_shift’:
numpy/core/src/umath/loops.c.src:830:18: internal compiler error: in expand_shift_1, at expmed.c:2349
             *out = in1 >> in2;
                  ^
numpy/core/src/umath/fast_loop_macros.h:180:9: note: in definition of macro ‘BASE_BINARY_LOOP_S_INP’
         op; \
         ^
numpy/core/src/umath/loops.c.src:828:5: note: in expansion of macro ‘BINARY_LOOP_FAST’
     BINARY_LOOP_FAST(@type@, @type@, {
     ^
Please submit a full bug report,
with preprocessed source if appropriate.
See <file:///usr/share/doc/gcc-5/README.Bugs> for instructions.
Running from numpy source directory.
```

**Review by seberg — APPROVED** [review]
Looks good to me. I have not worked with those really to be honest, but can't we put a new `npy_left_shift` to `numpy/core/src/npymath/npy_math_internal.h.src` or similar to avoid duplication?

**Comment by eric-wieser** [comment]
@seberg: We have no precedent in that file for functions templated over all integer types. The best we have are `lcm` and `gcd` which use the usual `l` and `ll` suffices. Unfortunately for shift we need to know the exact type, so we need overloads for short and byte as well.

What names would you propose?

**Comment by seberg** [comment]
Oh,had not checked, but doesn't sound too bad, the normal type code should suffice, e.g. `bhil` (and `q`, if we need `long long`) and the capital ones for unsigned of course? Most of these names never have underscoves, but I somewhat like extra underscores `npy_left_shift_@cc@` or `npy_lshift@cc@`, not sure I care too much (it should be private anyway).

**Comment by eric-wieser** [comment]
Realized that there's precedent for `h` and `hh` in printf specifiers, so the full set of suffices is `u?(h{0,2}|l{0,2})`

**Comment by eric-wieser** [comment]
Looks like this hit a problem on macOS as well. Somehow the FPU flag gets set, but only on computations with `int` and `unsigned int`.

File: numpy/core/src/umath/loops.c.src [hunk 1]
```diff
@@ -929,6 +929,43 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 @ATTR@ void
...
+NPY_NO_EXPORT NPY_GCC_OPT_3 void
+@TYPE@_left_shift@isa@(char **args, npy_intp *dimensions, npy_intp *steps,
+                  void *NPY_UNUSED(func))
+{
+    BINARY_LOOP_FAST(@type@, @type@, *out = npy_lshift@c@(in1, in2));
+
+#ifdef @TYPE@_left_shift_needs_clear_floatstatus
+    // For some reason, our macOS CI sets an "invalid" flag here, but only
+    // for some types.
+    npy_clear_floatstatus_barrier((char*)dimensions);
```
Comment (eric-wieser) [inline numpy/core/src/umath/loops.c.src]: I have no idea why this happens. Is our macOS run x86, or something else? Do we have a resident macOS expert?

**Comment by eric-wieser** [comment]
Coverage stats are just bad because we can't measure coverage for setup.py

File: numpy/core/src/umath/loops.c.src (hunk 1 above)
Comment (seberg) [reply numpy/core/src/umath/loops.c.src]: I have no idea, was thinking about putting it in godbold, but it is quite a lot of code around it (and I am not sure it is likely to show anything in any case).

@charris, I think I am good with this as is. But you had a look at it before and know `npy_math` stuff better. The floatstatus thing (and I guess the compile time check) are a bit unfortunate though...

File: numpy/core/src/npymath/npy_math_internal.h.src [hunk 2]
```diff
@@ -716,3 +716,44 @@ npy_@func@@c@(@type@ a, @type@ b)
...
+        return a << b;
+    }
+    else {
+        return 0;
+    }
+}
+NPY_INPLACE npy_@u@@type@
+npy_rshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
+{
+    if (NPY_LIKELY((size_t)b < sizeof(a) * CHAR_BIT)) {
```
Comment (charris) [inline numpy/core/src/npymath/npy_math_internal.h.src]: Do we want to deal with the possibility that `b` is negative?

Comment (charris) [reply numpy/core/src/npymath/npy_math_internal.h.src]: Not that currently the shift value is cast to unsigned and masked, at least on my hardware.
```
In [1]: a = array([1,2,3,4])                                                    

In [2]: a << -1                                                                 
Out[2]: 
array([-9223372036854775808,                    0, -9223372036854775808,
                          0])
```

Comment (seberg) [reply numpy/core/src/npymath/npy_math_internal.h.src]: I think there was a comment in the old/new code about that as well. Maybe we can push it off to another PR? That seems the easiest way forward? We can ignore the codecov failures.

Comment (mattip) [reply numpy/core/src/npymath/npy_math_internal.h.src]: xref 13898

**Comment by mattip** [comment]
Fixes #2249

**Comment by mattip** [comment]
With the caveat that negative shifts still need to be handled, I think this is good to go as is

File: numpy/core/src/umath/loops.c.src (hunk 1 above)
Comment (charris) [reply numpy/core/src/umath/loops.c.src]: Could be a clang thing.

**Comment by charris** [comment]
I'm OK with this going in. Would be good to have a release note about changed behavior in case it affects somebody downstream.

**Comment by mattip** [comment]
Waiting on resolution of #13908 to know how to add a release note

**Comment by mattip** [comment]
Since #13908 is taking a while, perhaps just add a not to `doc/release/1.18.0-notes.rst` and we will refactor that file if/when #13908 lands

**Comment by charris** [comment]
@seberg Ping.

**Comment by seberg** [comment]
@charris thanks. Let me merge the towncrier related PRs (since nobody complained), and then I will add a short release note to this. Other than that I remember this as being good (although that compiler bug is a bit annoying).

**Comment by seberg** [comment]
OK, I rebased this (it was a bit messy) and added a release note (we will have to modify the towncrier template a bit before the end, but it works).

Master seems to have added `@isa@` (and the GCC flag) since that exists on master, but I am not sure what the state of this was before the rebase? I do not think it had the `@isa@`, but also not sure how that adds up with the place in code. Is AVX worth it here?

File: numpy/core/src/npymath/npy_math_internal.h.src [hunk 3]
```diff
@@ -716,3 +716,44 @@ npy_@func@@c@(@type@ a, @type@ b)
...
+ * #type = byte, short, int, long, longlong#
+ * #c = hh,h,,l,ll#
+ */
+/**begin repeat1
+ *
+ * #u         = u,#
+ * #is_signed = 0,1#
+ */
+NPY_INPLACE npy_@u@@type@
+npy_lshift@u@@c@(npy_@u@@type@ a, npy_@u@@type@ b)
```
Comment (eric-wieser) [inline numpy/core/src/npymath/npy_math_internal.h.src]: Based on our learnings in #14121, it sounds like these need to end up in the header too.

Comment (eric-wieser) [reply numpy/core/src/npymath/npy_math_internal.h.src]: Done

**Comment by mattip** [comment]
Thanks @eric-wieser

**Pull request merged**